// @abp/ng.core/esm2022/lib/directives/init.directive.mjs
import { Directive, Output, EventEmitter, ElementRef } from '@angular/core';
import * as i0 from "@angular/core";
class InitDirective {
    constructor(elRef) {
        this.elRef = elRef;
        this.init = new EventEmitter();
    }
    ngAfterViewInit() {
        this.init.emit(this.elRef);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.6", ngImport: i0, type: InitDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.6", type: InitDirective, selector: "[abpInit]", outputs: { init: "abpInit" }, ngImport: i0 }); }
}
export { InitDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.6", ngImport: i0, type: InitDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[abpInit]' }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { init: [{
                type: Output,
                args: ['abpInit']
            }] } });
