// @abp/ng.core/esm2022/lib/directives/show-password.directive.mjs
import { Directive, ElementRef, Input, inject } from '@angular/core';
import * as i0 from "@angular/core";
class ShowPasswordDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
    }
    set abpShowPassword(visible) {
        const element = this.elementRef.nativeElement;
        if (!element)
            return;
        element.type = visible ? 'text' : 'password';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.6", ngImport: i0, type: ShowPasswordDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.6", type: ShowPasswordDirective, isStandalone: true, selector: "[abpShowPassword]", inputs: { abpShowPassword: "abpShowPassword" }, ngImport: i0 }); }
}
export { ShowPasswordDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.6", ngImport: i0, type: ShowPasswordDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[abpShowPassword]',
                }]
        }], propDecorators: { abpShowPassword: [{
                type: Input
            }] } });
